includes/class-wpcom-legacy-redirector.php

```diff
@@ -65,30 +66,26 @@ static function maybe_do_redirect() {
 			return;
 		}
 
-		$url = wp_parse_url( $_SERVER['REQUEST_URI'], PHP_URL_PATH );
-
-		if ( ! empty( $_SERVER['QUERY_STRING'] ) ) {
-			$url .= '?' . $_SERVER['QUERY_STRING'];
+		if ( ! isset( $_SERVER['REQUEST_URI'] ) ) {
+			return;
 		}
 
-		$request_path = apply_filters( 'wpcom_legacy_redirector_request_path', $url );
+		// $_SERVER is being sanitized in self::normalise_url
+		// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.ValidatedSanitizedInput.MissingUnslash
+		$redirect_data = Lookup::get_redirect_data( self::normalise_url( $_SERVER['REQUEST_URI'] ) );
 
-		if ( $request_path ) {
-			$redirect_uri = Lookup::get_redirect_uri( $request_path );
-			if ( $redirect_uri ) {
-				$redirect_status = apply_filters( 'wpcom_legacy_redirector_redirect_status', 301, $url );
-
-				// Third argument introduced to support the x_redirect_by header to denote WP redirect source.
-				if ( version_compare( get_bloginfo( 'version' ), '5.1.0', '>=' ) ) {
-					wp_safe_redirect( $redirect_uri, $redirect_status, WPCOM_LEGACY_REDIRECTOR_PLUGIN_NAME );
-				} else {
-					header( 'X-legacy-redirect: HIT' );
-					wp_safe_redirect( $redirect_uri, $redirect_status );
-				}
+		if ( ! $redirect_data || ! isset( $redirect_data['redirect_uri'] ) ) {
```

Comment:
It's OK to leave for now, but this line is now directly coupled into the keys used into the return of the `Lookup::get_redirect_data()` (specifically, `redirect_uri`).

Ideally, we'd have an object returned for that, which could have a suitable method on it which hid the implementation of the data structure.

Reply:
The `Lookup` class might actually need to be renamed to `Redirect` class as I see most of its methods would relate to that.

Reply:
I disagree - a `Redirect` is something that lives in the domain - it is an object that exists regardless of where the data for it is persisted. 

The Lookup, so far, is about the "concrete" implementation that says to look in the database.

Having an abstraction layer between the thing and where it lives, means we move towards a future that could store the redirects in a custom table, rather than the wp_posts table, or in memory for unit test purposes.